### PR TITLE
Milestone 1: Go module, shared packages, CI

### PR DESCRIPTION
**Pull request by rickeerickson**
## Summary

First step in migrating from shell/PowerShell scripts to Go CLIs.
This establishes the Go module, build tooling, CI pipeline, and
three shared `internal/` packages that both the Windows and Linux
CLIs will use. No CLI subcommands yet; those come in Milestone 2.

The three packages cover the core concerns shared across platforms:
leveled logging with audit-grade file output, command execution
with an allowlisted binary runner and Fibonacci backoff retry, and
YAML config loading with a full resolution chain (defaults, file,
env vars, CLI flags) and security-focused input validation.

## Changes

- **Go module**: `go.mod` with `github.com/rickeerickson/openwebui-wsl2-starter`
- **Build**: `Makefile` with build, cross-compile (linux/windows), test, lint, clean
- **CI**: GitHub Actions workflow (go vet, golangci-lint, go test -race, cross-compile, govulncheck)
- **Lint**: `.golangci.yml` v2 config with gosec, gocritic, bodyclose, noctx, errcheck
- **`internal/logging`**: Leveled logger (Error through Debug2), dual output (stderr + file), 0600 file permissions, injectable writer for tests
- **`internal/exec`**: Fibonacci backoff retry (pure functions), command runner with allowlisted binaries, context-cancellable retries
- **`internal/config`**: Config struct matching current `update_open-webui.config.sh` defaults, YAML loading via yaml.v3, viper-based resolution chain (defaults < file < env < flags), validation rejecting shell metacharacters in names
- **Repo hygiene**: Go patterns in `.gitignore`, diff drivers and binary patterns in `.gitattributes`

## Security

- Runner enforces an allowlist of 9 permitted binaries (NFR-SEC-2)
- All `exec.Command` calls use explicit argument lists, no shell interpolation (NFR-SEC-1)
- Config validation rejects `;`, `$`, backtick, and other shell metacharacters in container/volume/image names (NFR-SEC-3)
- All exec invocations are audit-logged with full argument lists (NFR-SEC-4)
- Log files written with 0600 permissions (NFR-SEC-5)

## Testing

61 unit tests across 3 packages, all passing with `-race`:
- `internal/logging`: 13 tests (30 subtests) covering level filtering, timestamp format, file permissions, VerbosityFromEnv edge cases, dual output, injectable writer
- `internal/exec`: 18 tests covering Fibonacci sequence, allowlist enforcement, retry exhaustion, context cancellation during sleep
- `internal/config`: 30 tests covering defaults, YAML loading, injection attempts (`;`, `$`, backtick), port boundaries, multi-error collection, full resolution chain precedence, round-trip marshal/unmarshal

Mutation testing via gremlins: 100% mutator coverage on all packages. Port boundary tests added to kill surviving mutants.

```
go test -race -count=1 ./...
# internal/config  1.652s
# internal/exec    1.320s
# internal/logging 1.477s

golangci-lint run ./...
# 0 issues

make build-all
# dist/ow-linux-amd64:       ELF 64-bit LSB executable, x86-64
# dist/ow-windows-amd64.exe: PE32+ executable (console) x86-64
```

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: internal/exec/runner_test.go [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+		t.Errorf("unexpected retry log: %s", buf.String())
+	}
+}
+
+func TestRunWithRetryRetriesOnFailure(t *testing.T) {
+	var buf bytes.Buffer
+	r := newTestRunner(t, &buf)
+
+	// sh -c script that fails twice then succeeds using a temp file as counter.
+	// We use RunWithRetry on "false" with 2 attempts to confirm retry happens.
```
[the remark this answers is not in the document]
Comment (rickeerickson) [reply internal/exec/runner_test.go]: Fixed in 32a3f2e.

File: Makefile [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+.PHONY: build build-linux build-windows build-all test lint clean
+
+build:
+	go build -ldflags "$(LDFLAGS)" -o dist/ow ./cmd/ow
+
+build-linux:
+	GOOS=linux GOARCH=amd64 go build -ldflags "$(LDFLAGS)" -o dist/ow-linux-amd64 ./cmd/ow
+
+build-windows:
```
[the remark this answers is not in the document]
Comment (rickeerickson) [reply Makefile:16]: Fixed in 32a3f2e. Added `@mkdir -p dist` to all build targets.

File: internal/exec/runner.go [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+	}
+
+	r.Logger.Info("exec: %s %v", name, args)
+
+	cmd := exec.CommandContext(ctx, name, args...) //nolint:gosec // G204: name is checked against allowlist above
+	out, err := cmd.CombinedOutput()
+	output := string(out)
+
+	r.Logger.Debug1("output: %s", output)
+
```
[the remark this answers is not in the document]
Comment (rickeerickson) [reply internal/exec/runner.go:49]: Fixed in 32a3f2e. Added NewRunner(logger) constructor that panics on nil.

File: internal/config/resolve.go [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+//  3. ~/.config/ow/ow.yaml (user-level)
+//  4. Environment variables (OW_ prefix)
+//  5. CLI flags (passed as map[string]string)
+//
+// Later sources override earlier ones.
+// Returns the resolved and validated config.
+//
+// Note: ollama.models cannot be set via environment variables because viper
+// handles string slices from env vars poorly. Models can only be set via
+// YAML file or flags.
```
[the remark this answers is not in the document]
Comment (rickeerickson) [reply internal/config/resolve.go]: Fixed in 32a3f2e. Updated comment to note neither env vars nor flags support models.

File: .github/workflows/go.yml [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+        with:
+          go-version-file: go.mod
+
+      - name: Shell and markdown checks
+        run: bash build_and_test.sh --shell --markdown
+
+      - run: go vet ./...
+
+      - name: Install golangci-lint
+        run: go install github.com/golangci/golangci-lint/v2/cmd/golangci-lint@latest
```
[the remark this answers is not in the document]
Comment (rickeerickson) [reply .github/workflows/go.yml]: Fixed in 32a3f2e. Pinned to v2.11.4 (matching local dev version).

File: internal/exec/runner.go [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+// RunWithRetry calls Run in a loop with Fibonacci backoff per RetryOpts.
+func (r *RealRunner) RunWithRetry(ctx context.Context, opts RetryOpts, name string, args ...string) (string, error) {
+	a := opts.InitialA
+	b := opts.InitialB
+
+	var lastOut string
+	var lastErr error
+
+	for attempt := 1; attempt <= opts.MaxAttempts; attempt++ {
```
[the remark this answers is not in the document]
Comment (rickeerickson) [reply internal/exec/runner.go:68]: Fixed in 32a3f2e. Added MaxAttempts < 1 validation with test.

File: .github/workflows/go.yml [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+
+      - run: golangci-lint run ./...
+
+      - run: go test -race -coverprofile=coverage.out ./...
+
+      - run: GOOS=linux GOARCH=amd64 go build -o /dev/null ./cmd/ow
+
+      - run: GOOS=windows GOARCH=amd64 go build -o /dev/null ./cmd/ow
+
+      - run: go install golang.org/x/vuln/cmd/govulncheck@latest && govulncheck ./...
```
[the remark this answers is not in the document]
Comment (rickeerickson) [reply .github/workflows/go.yml:35]: Intentionally kept at @latest. govulncheck is a security scanner; pinning means missing new vulnerability detections. Unlike a linter, new behavior here is the point.

File: internal/logging/logger.go [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+		return
+	}
+
+	msg := fmt.Sprintf(format, args...)
+	ts := time.Now().Format(timestampFormat)
+	line := fmt.Sprintf("%s - %s %s\n", ts, LevelString(level), msg)
+
+	_, _ = fmt.Fprint(l.stderr, line)
+	if l.logFile != nil {
+		_, _ = fmt.Fprint(l.logFile, line)
```
[the remark this answers is not in the document]
Comment (rickeerickson) [reply internal/logging/logger.go:77]: No action needed. The code already uses `_, _ = fmt.Fprint(...)` which satisfies errcheck. Lint passes clean both locally and in CI.

File: internal/logging/logger.go [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+// Close closes the log file, if one is open. Safe to call if no file is open.
+func (l *Logger) Close() error {
+	if l.logFile != nil {
+		return l.logFile.Close()
+	}
+	return nil
+}
+
+// LevelString returns a padded level name suitable for log output.
+// All strings are padded to match the width of "WARNING:" (8 chars + colon).
```
[the remark this answers is not in the document]
Comment (rickeerickson) [reply internal/logging/logger.go:115]: The comment says "8 chars wide" referring to the padded string width, not the literal character count of WARNING:. The padding logic is correct. Leaving as-is.

File: internal/exec/runner.go [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+
+		var delay time.Duration
+		delay, a, b = NextDelay(a, b)
+
+		r.Logger.Warn("retry %d/%d in %s: %s", attempt, opts.MaxAttempts, delay, lastErr)
+
+		select {
+		case <-ctx.Done():
+			return lastOut, fmt.Errorf("context cancelled during retry: %w", ctx.Err())
+		case <-time.After(delay):
```
[the remark this answers is not in the document]
Comment (rickeerickson) [reply internal/exec/runner.go:86]: This is a CLI tool that runs max 5 retries then exits. The timer is GC'd with the process. time.NewTimer + Stop adds complexity for zero practical benefit here. If this code moves to a long-lived server, worth revisiting.
